### PR TITLE
Paginate initial listings to display a small number of or records

### PR DESCRIPTION
**Pull request by TG1999**
Signed-off-by: Tushar Goel <tushar.goel.dav@gmail.com>

### CODE REVIEW

File: vulnerabilities/api.py [hunk 1]
```diff
@@ -175,9 +176,9 @@ def filter_purl(self, queryset, name, value):
 
 
 class PackageViewSet(viewsets.ReadOnlyModelViewSet):
-    queryset = Package.objects.all()
+    queryset = Package.objects.all().order_by("id")
     serializer_class = PackageSerializer
-    paginate_by = 50
+    pagination.PageNumberPagination.page_size = 10
```
Comment (tdruez) [inline vulnerabilities/api.py:181]: This seems quite wrong. It looks like you are overriding a value for a completely different class reference.

File: vulnerabilities/api.py [hunk 2]
```diff
@@ -175,9 +176,9 @@ def filter_purl(self, queryset, name, value):
 
 
 class PackageViewSet(viewsets.ReadOnlyModelViewSet):
-    queryset = Package.objects.all()
+    queryset = Package.objects.all().order_by("id")
```
Comment (tdruez) [inline vulnerabilities/api.py:179]: Why this change? If there is an issue with the ordering, please create a new issue and discuss the changes first.
This may need to be implement on the Model instead of on the API side. But the is no explanation about why you are changing all the ordering in the API context.

File: vulnerablecode/settings.py [hunk 3]
```diff
@@ -160,5 +160,5 @@
         "rest_framework.filters.SearchFilter",
     ),
     "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.PageNumberPagination",
-    "PAGE_SIZE": 100,
+    "PAGE_SIZE": 10,
```
Comment (tdruez) [inline vulnerablecode/settings.py:163]: Isn't this enough? The paginate_by on ViewSet is only useful to override the default setting, no?

**Review by tdruez — CHANGES_REQUESTED** [review]
The need for changing ordering in the context of refining the pagination is unclear.
See my inline comments.

Also, could you confirm that `bulk_search` actions is not limited to this 10 items pagination?
If I'm running a bulk search on 100 items, I do not want to have to make 10 requests to get the 100 items as results.

File: vulnerablecode/settings.py (hunk 3 above)
Comment (TG1999) [reply vulnerablecode/settings.py:163]: @tdruez yes, you are right, this is enough

File: vulnerabilities/api.py (hunk 2 above)
Comment (TG1999) [reply vulnerabilities/api.py:179]: @tdruez there is an issue reported here for same https://github.com/nexB/vulnerablecode/issues/685, but we can tackle this in a separate PR, will remove unrelated changes

Comment (tdruez) [reply vulnerabilities/api.py]: 1 issue -> 1 PR.

Also, this fix approach for #685 seems wrong.

Comment (TG1999) [reply vulnerabilities/api.py]: Alright! will fix this by changing in models side not at the API side

File: vulnerablecode/settings.py (hunk 3 above)
Comment (tdruez) [inline vulnerablecode/settings.py:163]: A comment about why we want to keep the default page size low would be useful.

Comment (TG1999) [reply vulnerablecode/settings.py:164]: Added the comment

File: vulnerablecode/settings.py [hunk 4]
```diff
@@ -160,5 +160,6 @@
         "rest_framework.filters.SearchFilter",
     ),
     "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.PageNumberPagination",
-    "PAGE_SIZE": 100,
+    # Reduce to 10 records per page to display data almost instantaneously https://github.com/nexB/vulnerablecode/issues/819
```
Comment (tdruez) [inline vulnerablecode/settings.py:163]: > almost instantaneously

This may not always be true, there are cases where displaying 10 records is not instantaneous.

Maybe something along:
Limit the load on the Database returning a small number of records by default.

**Review by tdruez — CHANGES_REQUESTED** [review]
Since we are already using the [PageNumberPagination](https://www.django-rest-framework.org/api-guide/pagination/#pagenumberpagination), I think we should enable the `page_size` feature, so the consumer can decide to get more items per requests when needed.
Let's keep a reasonable `max_page_size` though to avoid overloading the database.

File: vulnerabilities/api.py [hunk 5]
```diff
@@ -177,9 +183,9 @@ def filter_purl(self, queryset, name, value):
 class PackageViewSet(viewsets.ReadOnlyModelViewSet):
     queryset = Package.objects.all()
     serializer_class = PackageSerializer
-    paginate_by = 50
     filter_backends = (filters.DjangoFilterBackend,)
     filterset_class = PackageFilterSet
+    pagination_class = SmallResultSetPagination
```
Comment (tdruez) [inline vulnerabilities/api.py:188]: Why declaring this on the ViewSet instead of globally using the `DEFAULT_PAGINATION_CLASS` setting?

File: vulnerabilities/api.py [hunk 6]
```diff
@@ -24,6 +25,11 @@
 from vulnerabilities.models import VulnerabilitySeverity
 
 
+class SmallResultSetPagination(PageNumberPagination):
+    page_size_query_param = "page_size"
+    max_page_size = 30
```
Comment (tdruez) [inline vulnerabilities/api.py:30]: It could be useful to allow up to 100.

**Review by tdruez — CHANGES_REQUESTED** [review]

**Comment by tdruez** [comment]
@TG1999 We're missing an entry in the CHANGELOG about those changes.

**Review by tdruez — APPROVED** [review]
LGTM!

**Comment by TG1999** [comment]
Thanks! Merging
